### ISSUE TITLE
SocketExceptions in BinaryJedisPubSub

### ISSUE THREAD

**Issue by DivineTraube** [xetorthio/jedis#996]
Hi,

there is a non-trivial issue with the binary PubSub interface of Jedis. When issuing lots of subscribe (or unsubscribe) commands, Jedis randomly fails with Socket Exceptions.

Here is the code to reproduce this:
```java
Jedis jedis = new Jedis("localhost", 6379);
BinaryJedisPubSub listener = new BinaryJedisPubSub() {
};
CompletableFuture.runAsync(() -> jedis.subscribe(listener, SafeEncoder.encode("whatever")));
Thread.sleep(20);
for (int i = 0; i < 50000; i++) {
    listener.subscribe(SafeEncoder.encode(UUID.randomUUID().toString()));
}
```

The probability of crash gets higher as the number loop iterations increases.

```java
redis.clients.jedis.exceptions.JedisConnectionException: java.net.SocketException: Connection reset by peer: socket write error
	at redis.clients.jedis.Connection.flush(Connection.java:282)
	at redis.clients.jedis.BinaryJedisPubSub.subscribe(BinaryJedisPubSub.java:49)
	at info.orestes.events.redis.RedisEventsTest.testCrashPubSub(RedisEventsTest.java:136)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: java.net.SocketException: Connection reset by peer: socket write error
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at redis.clients.util.RedisOutputStream.flushBuffer(RedisOutputStream.java:31)
	at redis.clients.util.RedisOutputStream.flush(RedisOutputStream.java:213)
	at redis.clients.jedis.Connection.flush(Connection.java:279)
	... 29 more
```

On the network level, this happens because either the Redis server or the client (I've seen both) close the connection with a RST flag, with roughly 8000 in-flight unacknowledged bytes as can be seen with for instance Wireshark.  The strange thing is that the String-based JedisPubSub doesn't exhibit this behaviour. However, this is not a result of some poisoned byte sequence - even always sending the same byte produces an error. Furthermore the Redis protocol should allow arbitrary byte sequences as it only leverages the payload's length to determine its boundaries.

This also happens if all methods inside BinaryJedisPubSub are synchronized. It is also unrelated to the Redis Version and OS (tested on Mac OS and Windows w/ resp. newest Redis versions).

Any ideas on what causes this bug?

**Comment by marcosnils** [comment]
@DivineTraube nice finding. I could reproduce the issue with your example. 

I'll investigate later today hopefully to find the root cause of the problem. This is an issue where @itugs @HeartSaVioR may help

**Comment by marcosnils** [comment]
@DivineTraube I believe there's a race condition with the `runAsync()` or something. I don't know how it exactly works but I've changed the code as following and I'm not getting any errors.

```java
public static void main(String[] args) throws InterruptedException {
		Jedis jedis = new Jedis("localhost", 6379);

		final Boolean[] subscribed = {false};
		BinaryJedisPubSub listener = new BinaryJedisPubSub() {
			@Override
			public void onSubscribe(byte[] channel, int subscribedChannels) {
				if (!subscribed[0]) {
					subscribed[0] = true;
					System.out.println("Subscribed");
					for (int i = 0; i < 50000; i++) {
						this.subscribe(("channel"+i).getBytes());
					}
					System.out.println("Done");
				}

			}
		};
		jedis.subscribe(listener, SafeEncoder.encode("whatever"));
	}
```

**Comment by marcosnils** [comment]
@DivineTraube I've verified that `runAsync` internally uses a different thread and as you may know you can't use the same jedis instance in multiple threads as it causes unknown errors as the one you're seeing. You'll have to change your code by either accessing from the same thread (as my example) or use JedisPool for such purpose.

**Comment by DivineTraube** [comment]
@marcosnils I disagree. It's true that in the example the PubSub Listener runs in another thread. That is by design, since subscribe is a blocking call. The guide also says that subscribe may be called on a JedisPubSub instance. Though your example works, it is useless in practice - you cannot add and remove channels from an open PubSub connection, unless you happen to receive a message and can call subscribe/unsubscribe in the onMessage handler or ahead of time in the onSubscribe handler. If that is not the case, the PubSub Thread remains blocked on the socket, unable to issue subscribe calls as designed in the Redis protocol.

The pool is also not as solution, since it only provides new Jedis instances and hence new TCP connections. It can thus not be used to subscribe to additional channels in an already open connection. Using it would force you to open a new connection for every additional chanel/topic at runtime, which - apart from being inefficient (TCP slow start algorithm) - would very quickly exhaust not only the client but also the Redis server's capacities .

Imho the issue is still open: you cannot subscribe to additional channels at runtime without risking random socket errors.

**Comment by marcosnils** [comment]
@DivineTraube I agree with you that there's a limitation in Jedis where you can't "freely" subscribe to channels after the first `jedis.subscribe` (blocking) is executed. However my reason for closing  the issue is not the lack of that functionality which seems what you're highlighting. 

The reason why I closed the issues is because the title doesn't reflect the true discussion under this subject. The reason why your code was raising an exception is because the same `Jedis` instance is being accessed from multiple threads which is not supported by Jedis. 

Again, the fact that Jedis pub/sub implementation lacks some features and can be improved is something that should be opened in another issue  and labeled as en enhancement so we can propose new alternatives to this problem. I took the trouble to open a new issue #997 mentioning your observations where we can discuss further implementations. 

Thanks for making Jedis better, hopefully we can address this soon.

**Comment by xetorthio** [comment]
@DivineTraube I think we should propose how a new PubSub API should look like and see what we can do to get there.
I agree that the current API has huge limitations and is not really helpful. If you can, please help us to come up with a better API and implementation!
And as @marcosnils said, lets use #997 to design it :)

Thanks!

**Comment by xetorthio** [comment]
So I've been thinking about this a little more.
Why exactly are you subscribing to more channels immediately after subscribing to a few of them? Maybe the code you send is more of a way play code to reproduce the error you got.
But I am very curious to understand how exactly you are using JedisPubSub and why you need to subscribe to channels after subscribing for the first time.

**Comment by HeartSaVioR** [comment]
FYI, current issue can be resolved as #998, while we are still struggling about multi-threaded environment from #997.

**Comment by DivineTraube** [comment]
@xetorthio We have several use-cases. A particularly interesting one is project we'll release soon, where we use a set of wrapping interfaces to expose Redis as a [reliable broadcast](http://en.wikipedia.org/wiki/Distributed_algorithm#Standard_problems) channel with acknowledgements, durability and partition tolerance. It can be used by clients like a reliable pubsub with delivery guarantees.

Another example is a NoSQL student project that I teach. One team uses Redis (Cluster) for an XMPP chat application, where users/devices sign in at different frontend servers, that in turn use Pubsub to receive messages for an individual users over an individual channel. In order to scale, it is neccessary to have distinct channels for each user (not every server has to see all messages) and in order to be dynamic (e.g. load balancing, users signing on and off and registering), subscriptions change at runtime.

As for the API my favourite would be a fluent, asychronous interface like this:

```java
//Non-blockingly subscribe in the background (new Thread internally)
PubSub pubsub = PubSub.connect("localhost", 6379)
    .subscribe("firstChannel")
    .onMessage(handler) //handler should be a functional interface (channel, message) -> void, so Java 8 lambdas can be used
    .onSubscribe(otherHandler);

//later and blocking, potentially before connect is through
pubsub.subscribe("my", "other", "channels");
pubsub.subscribe("other");
```

Implementing this would be rather easy: all writing methods (subscribe, unsubscribe, etc.) have to be synchronized, so only one Thread at a time is able to write into the socket. This is sufficient as it guarantees atomic writes into the socket buffers, so there is no need to always use the *same* thread. Additionaly (which is now a manual pattern) the writing methods have to block on the connect using a Semaphore or CountdownLatch, so they don't interfere with the connection establishment.

@HeartSaVioR @marcosnils Thanks for investigating! Synchronization & #998 is workaround that works reasonably well until we have final solution.

**Comment by xetorthio** [comment]
@DivineTraube Regard your second example, which you gave more details. I am not sure how it works right now, but I remember implementing something similar a long time ago. And the way I did it was subscribing a user to a common "administration" channel and to a personal channel "user:id".
From the administration channel I get all kind of general notifications (messages to everyone, etc).
When the user is added to a new "room" it gets that notification from the administration channel, and subscribe to the new "room".
So I was using pubsub messages to trigger subscribes, from within the same thread.

**Comment by DivineTraube** [comment]
@xetorthio That might be workaround for some applications. But effectively it uses a broadcast network call to perform thread synchronization that could instead be implemented in the client through either synchronized socket writes or non-blocking I/O. Furthermore it is inefficient, because it spams non-related servers with control-information they don't need .
At the moment we are fine with fix #998 and synchronized subcribe and unsubscribe calls, which is the general-purpose solution.

**Comment by jordanglassman** [comment]
Using the example in https://github.com/xetorthio/jedis/issues/996#issue-79037922, jedis 2.9, redis 4.0.2, and with `i` bumped up to 5000000, I can't reproduce this issue anymore.
